### PR TITLE
ncdu: fix cpu impurity

### LINKED ISSUE #169461: ncdu_2: CPU features impurity
### Describe the bug
The output will include recent vector instructions if and only if the builder supports them. This break reproducibility.

### Steps To Reproduce
1. Get on an Intel Sandy Bridge machine
2. Run a derivation from a substituter. This results in a `SIGILL` due to a `vmovups xmm0 xmmword ptr` instruction
3. Build locally and run. No `SIGILL`

### Expected behavior
Both derivations should be identical.

### Additional context
This might affect `zig` projects in general. @andrewrk @AndersonTorres

This is both an impurity and a usability problem because I cannot use certain derivations on some of my older computers.

### Notify maintainers
@pSub @SuperSandro2000 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.34, NixOS, 22.05 (Quokka), 22.05.20220417.1ffba9f`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - nixpkgs: `/etc/nix/channels/nixos`
```

### LINKED ISSUE #185644: ncdu: Illegal instruction (core dumped)
### Describe the bug
The new version of ncdu crashes at runtime.
However, when run in a nix-shell, the 1.16 version is grabbed so it works:
```console
[gaetan@server:~]$ ncdu --version
ncdu 2.1.2

[gaetan@server:~]$ ncdu
Illegal instruction (core dumped)

[gaetan@server:~]$ nix-shell -p ncdu --command "ncdu --version"
ncdu 1.16

[gaetan@server:~]$ nix-shell -p ncdu --command ncdu       # It works
```

### Steps To Reproduce
1. Add `ncdu` to the system packages
2. run ncdu

### Expected behavior
The program works as expected.

### Additional context
I am using the `nixos-unstable` channel.
Bug report on ArchLinux: https://bugs.archlinux.org/task/74046

### Notify maintainers
@pSub @SuperSandro2000 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[gaetan@server:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.59, NixOS, 22.11 (Raccoon), 22.11.20220808.0e27b0d`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.10.3`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

### PR DESCRIPTION
**Pull request by ncfavier**
###### Description of changes

Fix https://github.com/NixOS/nixpkgs/issues/169461, fix https://github.com/NixOS/nixpkgs/issues/185644.

As discussed there and in https://github.com/NixOS/nixpkgs/issues/185665, a proper fix would be to add a setup hook that calls `zig build` with the right flags, but nobody has done this yet and I don't know zig well enough.

Tested by building on my laptop and running on my server: `SIGILL`s on master, works fine on this branch.

cc @lourkeur @andrewrk @GaetanLepage

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by ajs124** [comment]
While this fixes ncdu and we should probably merge this for now, IMO we need to fix this in zig somehow, e.g. by wrapping it or with a hook.
Otherwise we'll need to do this separately for every single zig application.

File: pkgs/tools/misc/ncdu/default.nix [hunk 1]
```diff
@@ -19,6 +19,8 @@ stdenv.mkDerivation rec {
 
   PREFIX = placeholder "out";
 
+  ZIG_FLAGS = "-Drelease-safe -Dcpu=baseline"; # see https://github.com/NixOS/nixpkgs/issues/169461
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/ncdu/default.nix:22]: Please also add a short comment why we do this in addition to the link.

Comment (ncfavier) [reply pkgs/tools/misc/ncdu/default.nix]: Done

**Review by SuperSandro2000 — APPROVED** [review]

**Review by trofi — APPROVED** [review]

**Review by pSub — APPROVED** [review]

File: pkgs/tools/misc/ncdu/default.nix [hunk 2]
```diff
@@ -19,6 +19,9 @@ stdenv.mkDerivation rec {
 
   PREFIX = placeholder "out";
 
+  # Avoid CPU feature impurity, see https://github.com/NixOS/nixpkgs/issues/169461
+  ZIG_FLAGS = "-Drelease-safe -Dcpu=baseline";
```
Comment (winterqt) [inline pkgs/tools/misc/ncdu/default.nix:23]: Why is release-safe required, instead of just setting the CPU to baseline?

Comment (ncfavier) [reply pkgs/tools/misc/ncdu/default.nix:23]: It's probably not required, but building with a "release" [mode](https://ziglang.org/documentation/master/#Build-Mode) seems reasonable for nixpkgs (mostly for reproducible builds), and `release-safe` is the one that's used [in most other places](https://github.com/NixOS/nixpkgs/issues/185665#issuecomment-1240555770) so I went with that one.

Which release mode to enable is a discussion that should be had when the zig setup hook is created.

**Pull request merged**